src/libraries/System.IO.Hashing/src/System/IO/Hashing/Crc32.X86.cs

```diff
@@ -0,0 +1,124 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Runtime.CompilerServices;
+using System.Runtime.Intrinsics.X86;
+using System.Runtime.Intrinsics;
+using System.Runtime.InteropServices;
+
+namespace System.IO.Hashing
+{
+    public partial class Crc32
+    {
+        private const int X86MinBufferSize = 64;
+
+        // Processes the bytes in source in X86MinBufferSize chunks using x86 intrinsics. After completion source is updated
+        // to refer to any remaining bytes (at most X86MinBufferSize-1). Requires support for Sse2 and Pclmulqdq intrinsics.
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static uint UpdateX86(uint crc, ref ReadOnlySpan<byte> source)
+        {
+            if (source.Length < X86MinBufferSize)
+            {
+                return crc;
+            }
+
+            // There's at least one block of 64.
+            Vector128<ulong> x1 = Vector128.Create(MemoryMarshal.Cast<byte, ulong>(source));
+            Vector128<ulong> x2 = Vector128.Create(MemoryMarshal.Cast<byte, ulong>(source.Slice(16)));
+            Vector128<ulong> x3 = Vector128.Create(MemoryMarshal.Cast<byte, ulong>(source.Slice(32)));
+            Vector128<ulong> x4 = Vector128.Create(MemoryMarshal.Cast<byte, ulong>(source.Slice(48)));
+            Vector128<ulong> x5;
+
+            x1 = Vector128.Xor(x1, Vector128.CreateScalar((ulong) crc));
+            Vector128<ulong> x0 = Vector128.Create(0x0154442bd4, 0x01c6e41596).AsUInt64(); // k1, k2
+
+            source = source.Slice(64);
+
+            // Parallel fold blocks of 64, if any.
+            while (source.Length >= 64)
+            {
+                x5 = Pclmulqdq.CarrylessMultiply(x1, x0, 0x00);
+                Vector128<ulong> x6 = Pclmulqdq.CarrylessMultiply(x2, x0, 0x00);
+                Vector128<ulong> x7 = Pclmulqdq.CarrylessMultiply(x3, x0, 0x00);
+                Vector128<ulong> x8 = Pclmulqdq.CarrylessMultiply(x4, x0, 0x00);
+
+                x1 = Pclmulqdq.CarrylessMultiply(x1, x0, 0x11);
+                x2 = Pclmulqdq.CarrylessMultiply(x2, x0, 0x11);
+                x3 = Pclmulqdq.CarrylessMultiply(x3, x0, 0x11);
+                x4 = Pclmulqdq.CarrylessMultiply(x4, x0, 0x11);
+
+                Vector128<ulong> y5 = Vector128.Create(MemoryMarshal.Cast<byte, ulong>(source));
+                Vector128<ulong> y6 = Vector128.Create(MemoryMarshal.Cast<byte, ulong>(source.Slice(16)));
+                Vector128<ulong> y7 = Vector128.Create(MemoryMarshal.Cast<byte, ulong>(source.Slice(32)));
+                Vector128<ulong> y8 = Vector128.Create(MemoryMarshal.Cast<byte, ulong>(source.Slice(48)));
```

Comment:
Similar to suggestion above. Work on `ref` instead of the ROS and use `LoadUnsafe` w/ the offset,
So to this method there could be passed the `ref` + `length` of the ROS, then also no `ref ROS` is needed as argument.

Reply:
I've completed this internally within the X86 method, but I'm still looking into the best option for moving away from `ref ROS` as a parameter. This method does advance the span and needs to return that to the caller so the caller can process remaining bytes dangling on the end. `ref ROS` is probably the cleanest way since you can't pass `ref ref byte` nor return `ref byte` in a `ValueTuple`.

Returning a private `ref struct` with a `ref` field is an option, but I'm not sure how well JIT will optimize that, `ref ROS` may be better. Another option may be to not take fall through to process the remaining bytes at all and process the remaining bytes directly in the X86 method.